### PR TITLE
Ensure that DoFs not touched by VectorTools::interpolate don't get set.

### LINKED ISSUE #7522: Problem with VectorTools::interpolate_based_on_material_id
From the mailing list in a post by Hamed Babaei:

> I would like to forcefully determine the solution at some regions of my domain in which I solve an initial value problem. Therefore, I chose to use VectorTools::interpolate_based_on_material_id function to implement the desired constant value for solution vector within particular regions for which I designate different material_ID. However, the problem is that despite the function description saying: 
> 
> "If a material id of some group of cells is missed in function_map, then dst will not be updated in the respective degrees of freedom of the output vector For example, if dst was successfully initialized to capture the degrees of freedom of the dof_handler of the problem with all zeros in it, then those zeros which correspond to the missed material ids will still remain in dst even after calling this function."
> 
> I receive -nan value for the cells whose material id is not included in function_map when using the interpolate_based_on_material_id function.  However, if I include all the material ids in the function_map it works as expected for the entire domain. This is not what I need because I would like to keep solution constant only for a small part of the domain, not the entire domain.

### LINKED ISSUE #7531: Shared Triangulation and Locally Owned Dofs issue
Hi everyone!
I believe there is some problem at least one of "locally_owned_dofs" and "map_dofs_to_support_points" on shared triangulation. The following test is a modified half of step-60, where I do the following:
- build a shared triangulation
- deform it
- ask for support points
- ask for the indices of locally relevant support points
- check if the support points with such indexes are really numbers, or contain -nan (I check only the first coordinate, but it is enough to trigger the problem).

There might be something wrong in my lengthy procedure, but with 1 process everything is fine (also by printing and checking manually the points), while with multiple cores it sometimes works, sometimes doesn't (depending on the deformation function and number of cores used)...

I am sorry if for the 250 lines; I have been trying to reduce it, but I couldn't understand how to produce a ```mesh_configuration_function``` without the parameters...(and, being a part of step-60, on the first run will fail and generate a .prm file )

```
#include <deal.II/base/logstream.h>
#include <deal.II/base/utilities.h>
#include <deal.II/base/timer.h>
#include <deal.II/base/parameter_acceptor.h>
#include <deal.II/grid/tria.h>
#include <deal.II/grid/grid_generator.h>
#include <deal.II/grid/grid_tools.h>
#include <deal.II/fe/fe.h>
#include <deal.II/fe/fe_q.h>
#include <deal.II/fe/fe_system.h>

#include <deal.II/fe/mapping_q_eulerian.h>
#include <deal.II/fe/mapping_fe_field.h>

#include <deal.II/dofs/dof_tools.h>
#include <deal.II/base/parsed_function.h>

#include <deal.II/distributed/grid_refinement.h>
#include <deal.II/distributed/solution_transfer.h>
#include <deal.II/distributed/tria.h>
#include <deal.II/distributed/shared_tria.h>

#include <deal.II/numerics/data_out.h>
#include <deal.II/numerics/vector_tools.h>
#include <deal.II/numerics/matrix_tools.h>
#include <deal.II/non_matching/coupling.h>

#include <deal.II/lac/affine_constraints.h>
#include <deal.II/lac/sparse_matrix.h>
#include <deal.II/lac/vector.h>
#include <deal.II/lac/linear_operator.h>
#include <deal.II/lac/linear_operator_tools.h>

#include <deal.II/base/mpi.h>

#include <math.h>
#include <iostream>
#include <fstream>

namespace Step60
{
  using namespace dealii;

  template <int dim, int spacedim = dim>
  class DistributedLagrangeProblem
  {
  public:
    class Parameters : public ParameterAcceptor
    {
    public:
      Parameters();

      unsigned int initial_mesh_refinement = 8;

      unsigned int embedding_space_finite_element_degree = 1;

      unsigned int mesh_space_finite_element_degree = 1;

      unsigned int mesh_configuration_finite_element_degree = 1;

      bool initialized = false;
    };

    DistributedLagrangeProblem(const Parameters &parameters);

    void run();

  private:
    // Object containing the actual parameters
    const Parameters &parameters;

    void setup_grids_and_dofs();

    void setup_mesh_dofs();

    std::unique_ptr< parallel::shared::Triangulation<dim,spacedim> > mesh_grid;
    std::unique_ptr<FiniteElement<dim, spacedim>> mesh_fe;
    std::unique_ptr<DoFHandler<dim, spacedim>>    mesh_dh;

    std::unique_ptr<FiniteElement<dim, spacedim>> mesh_configuration_fe;
    std::unique_ptr<DoFHandler<dim, spacedim>>    mesh_configuration_dh;
    Vector<double>                                mesh_configuration;

    ParameterAcceptorProxy<Functions::ParsedFunction<spacedim>>
      mesh_configuration_function;

    std::unique_ptr<Mapping<dim, spacedim>> mesh_mapping;

  };


  template <int dim, int spacedim>
  DistributedLagrangeProblem<dim, spacedim>::Parameters::Parameters()
    : ParameterAcceptor("/Distributed Lagrange<" +
                        Utilities::int_to_string(dim) + "," +
                        Utilities::int_to_string(spacedim) + ">/")
  {
    add_parameter("Initial embedded space refinement",
                  initial_mesh_refinement);

    add_parameter("Embedded space finite element degree",
                  mesh_space_finite_element_degree);

    add_parameter("Embedded configuration finite element degree",
                  mesh_configuration_finite_element_degree);

    parse_parameters_call_back.connect([&]() -> void { initialized = true; });
  }

  template <int dim, int spacedim>
  DistributedLagrangeProblem<dim, spacedim>::DistributedLagrangeProblem(
    const Parameters &parameters)
    : parameters(parameters)
    , mesh_configuration_function("Embedded configuration", spacedim)
  {
    mesh_configuration_function.declare_parameters_call_back.connect(
      []() -> void {
        ParameterAcceptor::prm.set("Function constants", "R=.3, Cx=.4, Cy=.4");


        ParameterAcceptor::prm.set("Function expression",
                                   "R*cos(2*pi*x)+Cx; R*sin(2*pi*x)+Cy");
      });
  }

  template <int dim, int spacedim>
  void DistributedLagrangeProblem<dim, spacedim>::setup_grids_and_dofs()
  {
    mesh_grid = std_cxx14::make_unique<parallel::shared::Triangulation<dim, spacedim>>(MPI_COMM_WORLD);
    GridGenerator::hyper_cube(*mesh_grid);
    mesh_grid->refine_global(parameters.initial_mesh_refinement);

    mesh_configuration_fe = std_cxx14::make_unique<FESystem<dim, spacedim>>(
      FE_Q<dim, spacedim>(
        parameters.mesh_configuration_finite_element_degree),
      spacedim);

    mesh_configuration_dh =
      std_cxx14::make_unique<DoFHandler<dim, spacedim>>(*mesh_grid);

    mesh_configuration_dh->distribute_dofs(*mesh_configuration_fe);
    mesh_configuration.reinit(mesh_configuration_dh->n_dofs());

    VectorTools::interpolate(*mesh_configuration_dh,
                             mesh_configuration_function,
                             mesh_configuration);

    mesh_mapping =
        std_cxx14::make_unique<MappingFEField<dim,
                                              spacedim,
                                              Vector<double>,
                                              DoFHandler<dim, spacedim>>>(
          *mesh_configuration_dh, mesh_configuration);

    setup_mesh_dofs();

    std::vector<Point<spacedim>> support_points(mesh_dh->n_dofs());
    DoFTools::map_dofs_to_support_points(*mesh_mapping,
                                           *mesh_dh,
                                           support_points);

    // We own only a part of the support points, those with the following index
    auto owned_dofs_idx = mesh_dh->locally_owned_dofs();

    // Cheking if the local support points are fine
    for(auto idx: owned_dofs_idx)
    {
      if( isnan(support_points[idx][0]) )
      {
        std::cout << "Non-valid point found: " << support_points[idx] << std::endl;
        AssertThrow(false,
                    ExcMessage("Sorry, this time not everything is fine... "
                               "A point containing NaN as coordinates is not valid! "
                               "Something is wrong! "));
      }
    }

    deallog << "Everything seems fine" << std::endl;
  }

  template <int dim, int spacedim>
  void DistributedLagrangeProblem<dim, spacedim>::setup_mesh_dofs()
  {
    mesh_dh =
      std_cxx14::make_unique<DoFHandler<dim, spacedim>>(*mesh_grid);
    mesh_fe = std_cxx14::make_unique<FE_Q<dim, spacedim>>(
      parameters.mesh_space_finite_element_degree);
    mesh_dh->distribute_dofs(*mesh_fe);
  }

  template <int dim, int spacedim>
  void DistributedLagrangeProblem<dim, spacedim>::run()
  {
    AssertThrow(parameters.initialized, ExcNotInitialized());
    deallog.depth_console(10);

    setup_grids_and_dofs();
  }
} // namespace Step60


int main(int argc, char **argv)
{
  try
    {
      using namespace dealii;
      using namespace Step60;
      auto mpi_initialization = Utilities::MPI::MPI_InitFinalize(argc, argv, 1);

      const unsigned int dim = 1, spacedim = 2;

      DistributedLagrangeProblem<dim, spacedim>::Parameters parameters;
      DistributedLagrangeProblem<dim, spacedim>             problem(parameters);

      std::string parameter_file;
      if (argc > 1)
        parameter_file = argv[1];
      else
        parameter_file = "parameters.prm";

      ParameterAcceptor::initialize(parameter_file, "used_parameters.prm");
      problem.run();
    }
  catch (std::exception &exc)
    {
      std::cerr << std::endl
                << std::endl
                << "----------------------------------------------------"
                << std::endl;
      std::cerr << "Exception on processing: " << std::endl
                << exc.what() << std::endl
                << "Aborting!" << std::endl
                << "----------------------------------------------------"
                << std::endl;
      return 1;
    }
  catch (...)
    {
      std::cerr << std::endl
                << std::endl
                << "----------------------------------------------------"
                << std::endl;
      std::cerr << "Unknown exception!" << std::endl
                << "Aborting!" << std::endl
                << "----------------------------------------------------"
                << std::endl;
      return 1;
    }
  return 0;
}
```

P.S. So, @luca-heltai , turns out I really found a problem...

### PR DESCRIPTION
**Pull request by bangerth**
Let's not merge this yet -- I think this fixes #7522, but I'd like to see what tests
this breaks before going further with it. If this doesn't break any tests, I'll
add a changelog and convert what is in #7522 to a proper testcase.

Fixes #7531.

edit (masterleinad): Link #7531.

### CODE REVIEW

**Comment by masterleinad** [comment]
I believe we were discussing this during the last workshop with @GivAlz but I can't find the PR or issue right now.

**Comment by GivAlz** [comment]
Yes, same issue, the PR is this https://github.com/dealii/dealii/pull/6964/files , I thought there was something wrong with changing such a basic function (so I let the PR hanging forever and closed it yesterday...), but it turns out it was the right...I am deeply sorry about this :( Thank you Bangerth!

**Comment by bangerth** [comment]
Ah, how interesting -- we started from completely different problems and arrived at the same solution...

**Comment by bangerth** [comment]
I'll add the test from #6964 as well then.

**Comment by GivAlz** [comment]
> I'll add the test from #6964 as well then.

Glad at least that can be recovered :)

**Comment by masterleinad** [comment]
> I'd like to see what tests this breaks before going further with it.

The test suite passes with my configuration.

**Comment by masterleinad** [comment]
Just as with #6964, this PR makes sense to me.

**Comment by bangerth** [comment]
I ran out of time last week to finish this, and I'm mostly going to be offline this week. I'll get back to it after Christmas.

File: tests/vector_tools/interpolate_with_material_id_01.cc [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+
+
+// Check that interpolate_based_on_material_id() really does ignore
+// material_ids not touch cells not listed
```
Comment (masterleinad) [inline tests/vector_tools/interpolate_with_material_id_01.cc]:
```suggestion
// cells with a material_id not included in the provided std::map.
```

File: tests/vector_tools/interpolate_with_material_id_01.cc [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+#include <deal.II/grid/tria_iterator.h>
+
+#include <deal.II/lac/dynamic_sparsity_pattern.h>
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/solver_cg.h>
+#include <deal.II/lac/sparsity_tools.h>
+#include <deal.II/lac/trilinos_precondition.h>
+#include <deal.II/lac/trilinos_solver.h>
+#include <deal.II/lac/trilinos_sparse_matrix.h>
+#include <deal.II/lac/trilinos_vector.h>
```
Comment (masterleinad) [inline tests/vector_tools/interpolate_with_material_id_01.cc]: We are not using `Trilinos` in this test.

File: tests/vector_tools/interpolate_with_material_id_01.cc [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+  ///////////////////////////////////////////////////////////////
+
+  template <int dim>
+  class Solid
+  {
+  public:
+    Solid();
+
+    virtual ~Solid();
```
Comment (masterleinad) [inline tests/vector_tools/interpolate_with_material_id_01.cc]: We don't need a `virtual` destructor here.

**Comment by bangerth** [comment]
@masterleinad: I fixed the places you've marked up!

**Pull request merged**